stackrox/stackrox
# ROX-18957: Delete report configurations that have invalid scopeIDs

## Description

In 4.0 (as part of Collections), report configurations were auto-migrated to use collections. In rare cases, some of their scopes couldn't be migrated and they were left in a broken state (with notice to users). As pointed out in the [4.1 release notes](https://docs.openshift.com/acs/4.1/release_notes/41-release-notes.html#notice-upcoming-report-change_release-notes-41), we now want to delete it in 4.2.

There are two columns that have collection ids. `scopeID` and `collectionID`. The latter is new and not yet released so should be empty. This migration deletes all reports where (`scopeID` is empty AND `collectionID` is empty) OR (`scopeID` points to a non-existent collection record). 

## Checklist
~[ ] Determined and documented upgrade steps

If any of these don't apply, please comment below.

## Testing Performed

Automated tests.

Manual tests:

1. Created a bunch of reports, manually edited DB to delete or change collection id and then upgraded.
2. Created the following scope in 3.74:
```
{
    "id": "",
    "name": "Scope cannot be converted to collection",
    "description": "",
    "rules":
    {
        "clusterLabelSelectors":
        [
            {
                "requirements":
                [
                    {
                        "key": "key1",
                        "op": "NOT_IN",
                        "values":
                        [
                            "val1",
                            "val2"
                        ]
                    }
                ]
            }
        ]
    }
}
```

Updated to 4.0 and noticed this in logs:

```
migrations/m_171_to_m_172_move_scope_to_collection_in_report_configurations: 2023/08/17 01:32:58.081881 migration.go:196: Error: Failed to create collections for scope <"Scope cannot be converted to collection">: Unsupported operator NOT_IN in scope's label selectors. Only operator 'IN' is supported.
 The scope is attached to the following report configurations: [invalid report]; Please manually create an equivalent collection and edit the listed report configurations to use this collection. Note that reports will not function correctly until a collection is attached.
```

Verified that the report existed with an error for collection

Then upgraded to this image.

Verified this in logs:
```
migrations/m_187_to_m_188_remove_reports_without_migrated_collections: 2023/08/17 01:34:14.971207 migration_impl.go:34: Info: Deleted 1 report configurations because they have invalid collections
```

And visually verified that the report was deleted

In addition to reviewing your code, reviewers **must** also review your testing
instructions and make sure they are sufficient.
